src/data_structures/annot_map.rs

```diff
@@ -0,0 +1,299 @@
+//! Efficient container for locations annotated across a set of named
+//! reference sequences.
+use std::collections::HashMap;
+use std::hash::Hash;
+
+use bio_types::annot::loc::Loc;
+use data_structures::interval_tree;
+use data_structures::interval_tree::{IntervalTree, IntervalTreeIterator};
+use utils::Interval;
+
+/// Efficient container for querying annotations, using `HashMap` and
+/// `IntervalTree`.
+///
+/// The container is parameterized over the type of the reference
+/// sequence names `R` (which is often a `String`) and the type of the
+/// contained objects `T`.
+///
+/// The container finds annotations that overlap a specific query
+/// location. Overlaps are identified without regard for strandedness
+/// and without regard for e.g. spliced-out introns within the
+/// annotation or the query.
+///
+/// Thus, the overlapping annotations identified by querying a
+/// `AnnotMap` may need further filtering.
+#[derive(Debug, Clone)]
+pub struct AnnotMap<R, T>
+where
+    R: Hash + Eq,
+{
+    refid_itrees: HashMap<R, IntervalTree<isize, T>>,
+}
+
+impl<R, T> AnnotMap<R, T>
+where
+    R: Eq + Hash,
+{
+    /// Creates a new, empty `AnnotMap`.
+    ///
+    /// ```
+    /// extern crate bio;
+    /// use bio::data_structures::annot_map::AnnotMap;
+    /// let mut genes: AnnotMap<String,String> = AnnotMap::new();
+    /// ```
+    pub fn new() -> Self {
+        AnnotMap {
+            refid_itrees: HashMap::new(),
+        }
+    }
+
+    /// Inserts an object into the container at a specified location.
+    ///
+    /// ```
+    /// extern crate bio_types;
+    /// extern crate bio;
+    /// use bio_types::strand::ReqStrand;
+    /// use bio_types::annot::contig::Contig;
+    /// use bio::data_structures::annot_map::AnnotMap;
+    /// let mut genes: AnnotMap<String,String> = AnnotMap::new();
+    /// let tma22 = Contig::new("chrX".to_owned(), 461829, 462426 - 461829, ReqStrand::Forward);
+    /// genes.insert_at("TMA22".to_owned(), &tma22);
+    /// let tma19 = Contig::new("chrXI".to_owned(), 334412, (334916 - 334412), ReqStrand::Reverse);
+    /// genes.insert_at("TMA19".to_owned(), &tma19);
+    /// ```
+    pub fn insert_at<L>(&mut self, t: T, l: &L)
```

Comment:
Please give the arguments more speaking names.

Reply:
Good point, I've changed these and described the arguments in doc comments.